File: CaveEscape/src/CaveEscape8.pas
program GameMain;
uses SwinGame, sgTypes, sgTimers, sgSprites, sysUtils;

const
  GRAVITY = 0.08;
  JUMP_RECOVERY_BOOST = 2;
  MAX_SPEED = 5;
  FOREGROUND_FOREROOF_POLE_SCROLL_SPEED = -2;
  BACKGROUND_SCROLL_SPEED = -1;

type
  PoleData = record
    ScoreLimiter: Boolean;
    UpPole: Sprite;
    DownPole: Sprite;
  end;

  Poles = array [0..3] of PoleData;

  GameData = record
    Player: Sprite;
    Foreroof: Sprite;
    Foreground: Sprite;
    Background: Sprite;
    IsDead: Boolean;
    Score: Integer;
    Poles: Poles;
    PoleReleaseDistance: Integer;
  end;

function GetRandomPoles(): PoleData;
begin
    result.UpPole := CreateSprite(BitmapNamed('UpPole'));
    result.DownPole := CreateSprite(BitmapNamed('DownPole'));
    SpriteSetX(result.UpPole, ScreenWidth());
    SpriteSetY(result.UpPole, ScreenHeight() - SpriteHeight(result.UpPole) - RND(BitmapHeight(BitmapNamed('Foreground'))));
    SpriteSetX(result.DownPole, SpriteX(result.UpPole));
    SpriteSetY(result.DownPole, RND(BitmapHeight(BitmapNamed('Foreroof'))));
    SpriteSetDx(result.UpPole, 0);
    SpriteSetDx(result.DownPole, 0);
    result.ScoreLimiter := true;
end;

function GetNewPlayer(): Sprite;
begin
  result := CreateSprite(BitmapNamed('Player'), AnimationScriptNamed('PlayerAnimations'));
  SpriteSetX(result, ScreenWidth() / 2 - SpriteWidth(result));
  SpriteSetY(result, ScreenHeight() / 2);
  SpriteStartAnimation(result, 'Fly');
end;

procedure SetUpBackground(var Background, Foreground, Foreroof: Sprite);
begin
  Background := CreateSprite(BitmapNamed('Background'));
  SpriteSetX(Background, 0);
  SpriteSetY(Background, 0);
  SpriteSetDx(Background, BACKGROUND_SCROLL_SPEED);

  Foreground := CreateSprite(BitmapNamed('Foreground'), AnimationScriptNamed('ForegroundAminations'));
  SpriteSetX(Foreground, 0);
  SpriteSetY(Foreground, ScreenHeight() - SpriteHeight(Foreground));
  SpriteSetDx(Foreground, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  Foreroof := CreateSprite(BitmapNamed('Foreroof'));
  SpriteSetX(Foreroof, 0);
  SpriteSetY(Foreroof, 0);
  SpriteSetDx(Foreroof, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
end;

procedure SetUpGame(var gData: GameData);
var
  i: Integer;
begin
  LoadResourceBundleNamed('CaveEscape', 'CaveEscape.txt', false);

  for i:= Low(gData.Poles) to High(gData.Poles) do
  begin
    gData.Poles[i] := GetRandomPoles();
  end;
  gData.player := GetNewPlayer();
  gData.Score := 0;
  gData.IsDead := false;
  gData.PoleReleaseDistance := 0;
  SetUpBackground(gData.Background, gData.Foreground, gData.Foreroof);

  SpriteStartAnimation(gData.Foreground, 'Fire');
end;

procedure UpdateVelocity(var toUpdate: Sprite);
begin
  SpriteSetDy(toUpdate, SpriteDy(toUpdate) + GRAVITY);

  if SpriteDy(toUpdate) > MAX_SPEED then
  begin
    SpriteSetDy(toUpdate, MAX_SPEED);
  end
  else if SpriteDy(toUpdate) < -(MAX_SPEED) then
  begin
    SpriteSetDy(toUpdate, -(MAX_SPEED));
  end;
end;

procedure UpdateBackground(var gData: GameData);
begin
  UpdateSprite(gData.foreGround);
  UpdateSprite(gData.Foreroof);
  updateSprite(gData.Background);
  if (SpriteX(gdata.Foreground) <= -(SpriteWidth(gData.ForeGround) / 2)) then
  begin
    SpriteSetX(gData.Foreground, 0);
    SpriteSetX(gData.Foreroof, 0);
  end;
  if (SpriteX(gdata.Background) <= -(SpriteWidth(gData.Background) / 2)) then
  begin
    SpriteSetX(gData.Background, 0);
  end;
end;

procedure CheckForCollisions(var toUpdate: GameData);
var
  i: Integer;
begin
  if (SpriteCollision(toUpdate.player, toUpdate.Foreground)) or (SpriteCollision(toUpdate.player, toUpdate.Foreroof)) then
  begin
    toUpdate.IsDead := true;
    exit;
  end;

  for i := Low(toUpdate.Poles) to High(toUpdate.Poles) do
  begin
    if SpriteCollision(toUpdate.player, toUpdate.Poles[i].UpPole) or SpriteCollision(toUpdate.player, toUpdate.Poles[i].DownPole)then
    begin
      toUpdate.IsDead := true;
      exit;
    end;
  end;
end;

procedure UpdatePlayer(var toUpdate: Sprite);
begin
  UpdateVelocity(toUpdate);
  UpdateSprite(toUpdate);
end;

procedure HandleInput(var toUpdate: Sprite);
begin
  if KeyTyped(SpaceKey) then
  begin
    SpriteSetDy(toUpdate, SpriteDy(toUpdate) + -(JUMP_RECOVERY_BOOST));
  end;
end;

procedure ResetPoleData(var toReset: PoleData);
begin
  FreeSprite(toReset.UpPole);
  FreeSprite(toReset.DownPole);
  toReset := GetRandomPoles();
end;

procedure MarkPoleForMovement(var poles: Poles; var releaseDistance: Integer);
var
  i: Integer;
begin
  releaseDistance += FOREGROUND_FOREROOF_POLE_SCROLL_SPEED;
  if releaseDistance <= 0 then
  begin
    for i := Low(poles) to High(poles) do
    begin
      if SpriteDx(poles[i].UpPole) = 0 then
      begin
        SpriteSetDx(poles[i].UpPole, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
        SpriteSetDx(poles[i].DownPole, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
        releaseDistance := RND(SpriteWidth(poles[0].UpPole)) + SpriteWidth(poles[0].UpPole);
        break;
      end;
    end;
  end;
end;

procedure UpdatePoles(var myGame: GameData);
var
  i: Integer;
begin
  MarkPoleForMovement(myGame.poles, myGame.PoleReleaseDistance);
  for i:= Low(myGame.Poles) to High(myGame.Poles) do
  begin
    UpdateSprite(myGame.Poles[i].UpPole);
    UpdateSprite(myGame.Poles[i].DownPole);

    if SpriteX (myGame.Poles[i].UpPole) < (SpriteX(myGame.player)) then
    begin
      if (myGame.Poles[i].ScoreLimiter = true) then
      begin
        myGame.Poles[i].ScoreLimiter := false;
        myGame.Score += 1;
      end;
    end;

    if (SpriteOffscreen(myGame.Poles[i].UpPole)) and (SpriteOffscreen(myGame.Poles[i].DownPole) and (myGame.Poles[i].ScoreLimiter = false))then
    begin
      ResetPoleData(myGame.Poles[i]);
    end;
  end;
end;

procedure ResetGame(var gData: GameData);
var
  i: Integer;
begin
  gData.Player := GetNewPlayer();
  for i:= Low(gData.Poles) to High(gData.Poles) do
  begin
    ResetPoleData(gData.Poles[i]);
  end;
  gData.IsDead := false;
  gData.Score := 0;
end;

procedure UpdateGame(var gData: GameData);
begin
  if not (gData.IsDead) then
  begin
    CheckForCollisions(gData);
    HandleInput(gData.player);
    UpdateBackground(gdata);
    UpdatePlayer(gData.player);
    UpdatePoles(gData);
  end
  else //The player has died :(
  begin
    ResetGame(gData);
  end;
end;

procedure DrawPoles(const myPoles: Poles);
var
  i: Integer;
begin
  for i:= Low(myPoles) to High(myPoles) do
  begin
    DrawSprite(myPoles[i].UpPole);
    DrawSprite(myPoles[i].DownPole);
  end;
end;

procedure DrawGame(const gData: GameData);
begin
  DrawSprite(gData.Background);
  DrawPoles(gData.Poles);
  DrawSprite(gData.Foreroof);
  DrawSprite(gData.ForeGround);
  DrawSprite(gData.player);
  DrawText(IntToStr(gData.score), ColorWhite, 'GameFont', 10, 0);
end;

procedure Main();
var
  gData: GameData;
begin
  OpenGraphicsWindow('Cave Escape', 432, 768);
  OpenAudio();
  SetUpGame(gData);

  FadeMusicIn('GameMusic', -1, 15000);

  repeat // The game loop...
    ProcessEvents();
    ClearScreen(ColorWhite);
    UpdateGame(gData);
    DrawGame(gData);
    RefreshScreen();
  until WindowCloseRequested();
end;

begin
  Main();
end.

File: C/CaveEscape/src/cave_escape_8.cpp
#include <stdio.h>
#include "SwinGame.h"

#define GRAVITY 0.08
#define MAX_SPEED 5
#define JUMP_RECOVERY_BOOST 2
#define FOREGROUND_FOREROOF_POLE_SCROLL_SPEED -2
#define BACKGROUND_SCROLL_SPEED -1
#define NUM_POLES 4

typedef struct pole_data
{
  sprite up_pole, down_pole;
} pole_data;

typedef struct pole_data poles[NUM_POLES];

typedef struct background_data
{
  sprite foreroof, foreground, background;
} background_data;

typedef struct game_data
{
  sprite player;
  background_data scene;
  poles poles;
} game_data;

sprite get_new_player()
{
  sprite result;

  result = create_sprite(bitmap_named("Player"), animation_script_named("PlayerAnimations"));
  sprite_set_x(result, screen_width() / 2 - sprite_width(result));
  sprite_set_y(result, screen_height() / 2);
  sprite_start_animation(result, "Fly");

  return result;
}

pole_data get_random_poles()
{
  pole_data result;

  result.up_pole = create_sprite(bitmap_named("UpPole"));
  result.down_pole = create_sprite(bitmap_named("DownPole"));
  sprite_set_x(result.up_pole, screen_width() + rnd(1200));
  sprite_set_y(result.up_pole, screen_height() - sprite_height(result.up_pole));
  sprite_set_x(result.down_pole, sprite_x(result.up_pole));
  sprite_set_y(result.down_pole, 0);
  sprite_set_dx(result.up_pole, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  sprite_set_dx(result.down_pole, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);

  return result;
}

background_data get_new_background()
{
  background_data result;

  result.background = create_sprite(bitmap_named("Background"));
  sprite_set_x(result.background, 0);
  sprite_set_y(result.background, 0);
  sprite_set_dx(result.background, BACKGROUND_SCROLL_SPEED);

  result.foreground = create_sprite(bitmap_named("Foreground"), animation_script_named("ForegroundAminations"));
  sprite_set_x(result.foreground, 0);
  sprite_set_y(result.foreground, screen_height() - sprite_height(result.foreground));
  sprite_set_dx(result.foreground, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  sprite_start_animation(result.foreground, "Fire");

  result.foreroof = create_sprite(bitmap_named("Foreroof"));
  sprite_set_x(result.foreroof, 0);
  sprite_set_y(result.foreroof, 0);
  sprite_set_dx(result.foreroof, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);

  return result;
}

void handle_input(sprite player)
{
  if (key_typed(SPACE_KEY))
  {
    sprite_set_dy(player, sprite_dy(player) - JUMP_RECOVERY_BOOST);
  }
}

void reset_pole_data(pole_data poles)
{
  free_sprite(poles.up_pole);
  free_sprite(poles.down_pole);
  poles = get_random_poles();
}

void update_velocity(sprite player)
{
  sprite_set_dy(player, sprite_dy(player) + GRAVITY);

  if (sprite_dy(player) > MAX_SPEED)
  {
    sprite_set_dy(player, MAX_SPEED);
  }
  else if (sprite_dy(player) < -(MAX_SPEED))
  {
    sprite_set_dy(player, -(MAX_SPEED));
  }
}

void update_poles(pole_data poles)
{
  update_sprite(poles.up_pole);
  update_sprite(poles.down_pole);

  if ((sprite_x(poles.up_pole) + sprite_width(poles.up_pole) < 0) && (sprite_x(poles.down_pole) + sprite_width(poles.down_pole) < 0))
  {
    reset_pole_data(poles);
  }
}

void update_poles_array(poles poles_array)
{
  int i;

  for (i = 0; i < NUM_POLES; i++)
  {
    update_poles(poles_array[i]);
  }
}

void update_background(background_data *scene)
{
  update_sprite(scene->foreground);
  update_sprite(scene->foreroof);
  update_sprite(scene->background);

  if (sprite_x(scene->foreground) <= -(sprite_width(scene->foreground) / 2))
  {
    sprite_set_x(scene->foreground, 0);
    sprite_set_x(scene->foreroof, 0);
  }
  if (sprite_x(scene->background) <= -(sprite_width(scene->background) / 2))
  {
    sprite_set_x(scene->background, 0);
  }
}

void update_player(sprite player)
{
  update_velocity(player);
  update_sprite(player);
}

void update_game(game_data* game)
{
  handle_input(game->player);
  update_background(&game->scene);
  update_player(game->player);
  update_poles_array(game->poles);
}

void draw_game(game_data* game)
{
  int i;

  draw_sprite(game->scene.background);
  for (i = 0; i < NUM_POLES; i++)
  {
    draw_sprite(game->poles[i].up_pole);
    draw_sprite(game->poles[i].down_pole);
  }
  draw_sprite(game->scene.foreroof);
  draw_sprite(game->scene.foreground);
  draw_sprite(game->player);
}

void set_up_game(game_data* game)
{
  int i;

  load_resource_bundle_named("CaveEscape", "CaveEscape.txt", false);
  for (i = 0; i < NUM_POLES; i++)
  {
    game->poles[i] = get_random_poles();
  }
  game->player = get_new_player();
  game->scene = get_new_background();
}

int main()
{
    game_data game;

    open_graphics_window("Cave Escape", 432, 768);
    set_up_game(&game);

    do
    {
      process_events();
      clear_screen(ColorWhite);
      update_game(&game);
      draw_game(&game);
      refresh_screen();

    } while(!window_close_requested());

    return 0;
}
